src/pubsub/ua_pubsub.c

```diff
@@ -182,6 +185,1098 @@ UA_Server_removeWriterGroup(UA_Server *server, const UA_NodeId writerGroup){
+    UA_PubSubDataSetReader *dataSetReader, *tmpDataSetReader;
+    LIST_FOREACH_SAFE(dataSetReader, &readerGroup->readers, listEntry, tmpDataSetReader) {
+        UA_PubSubDataSetReader_delete(server, dataSetReader);
+    }
+    UA_PubSubConnection* pConn = UA_PubSubConnection_findConnectionbyId(server, readerGroup->linkedConnection);
+    if (pConn != NULL) {
+        pConn->readerGroupsSize--;
+    }
+
+    //Delete ReaderGroup and its members
+    UA_String_deleteMembers(&readerGroup->config.name);
+    UA_NodeId_deleteMembers(&readerGroup->linkedConnection);
+    UA_NodeId_deleteMembers(&readerGroup->identifier);
+}
+
+/**
+ * Copy ReaderGroup configuration.
+ *
+ * @param source
+ * @param destination
+ * @return UA_STATUSCODE_GOOD on success
+ */
+UA_StatusCode
+UA_PubSubReaderGroupConfig_copy(const UA_PubSubReaderGroupConfig *src,
+                                UA_PubSubReaderGroupConfig *dst) {
+    UA_String_copy(&src->name, &dst->name);
+    //Currently simple memcpy only
+    memcpy(&dst->securityParameters, &src->securityParameters, sizeof(UA_PubSubSecurityParameters));
+    return UA_STATUSCODE_GOOD;
+}
+
+/**
+ * Process NetworkMessage.
+ *
+ * @param server
+ * @param networkmessage
+ * @return UA_STATUSCODE_GOOD on success
+ */
+UA_StatusCode
+UA_PubSubConnection_processNetworkMessage(UA_Server *server, UA_NetworkMessage *pMsg,
+                                          UA_PubSubConnection *pConnection) {
+    UA_StatusCode retval = UA_STATUSCODE_BADNOTIMPLEMENTED;
+    if ((pMsg == NULL) || (pConnection == NULL)) {
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+    }
+
+    UA_Boolean publisherIdAvailabe = false;
+    if (pMsg->publisherIdEnabled) {
+        publisherIdAvailabe = true;
+    }
+
+    UA_Boolean writerGroupIdAvailabe = false;
+    if (pMsg->groupHeaderEnabled && pMsg->groupHeader.writerGroupIdEnabled) {
+        writerGroupIdAvailabe = true;
+    }
+
+    UA_Boolean dataSetWriterIdAvailabe = false;
+    if (pMsg->payloadHeaderEnabled) {
+        dataSetWriterIdAvailabe = true;
+    }
+
+    // DataSetClassId is not necessarily able to identify the corresponding DataSetReader:
+    // all DataSetMessages in the msg have the same DataSetClassId. Anyway it should be checked
+    UA_Boolean dataSetClassIdAvailable = false;
+    if (pMsg->dataSetClassIdEnabled) {
+        dataSetClassIdAvailable = true;
+    }
+
+    // ToDo: Here some filtering is possible.
+    UA_Byte anzDataSets = 1;
+    if (pMsg->payloadHeaderEnabled) {
+        anzDataSets = pMsg->payloadHeader.dataSetPayloadHeader.count;
+    }
+
+    for (UA_Byte iterator = 0; iterator < anzDataSets; iterator++) {
+        UA_PubSubDataSetReader* dataSetReaderErg = NULL;
+        if (publisherIdAvailabe || writerGroupIdAvailabe || dataSetWriterIdAvailabe) {
+            UA_PubSubReaderGroup* readerGroup = NULL;
+            LIST_FOREACH(readerGroup, &pConnection->readerGroups, listEntry) {
+                UA_PubSubDataSetReader *tmpReader;
```

Comment:
Here we have nested code with:

- three for-loops
- four if-statements
- a switch statement

That is way too complex.
I don't see any bugs. But I can also not see the absence of bugs. ;-)

Factor out a dedicated method like "getDataSetReadByAttribute" or whatever the correct name should be.